Add encryption support

q3st1on opened the pull request:
# Adding Encryption Support
As demonstrated by Rizzo and Duong<sup>[1]</sup> padding oracles in CBC encryption implementations can be used to form an encryption oracle, allowing an attacker to forge valid ciphertexts without knowing the secret key. This PR contains a rather quick implementation of the CBC-R attack, tests for this encryption mode and an updated README file. The changes are not breaking, with the default behaviour remaining the original decryption routine. Wrapping functions have been duplicated for the encryption routine to allow for modification to better suite it in future, but this is beyond what I need right now and as such they have been left unchanged.

### The CBC-R attack:
for a padded plaintext we want to encrypt, we divide it into _n_ blocks and _n+1_ random ciphertext blocks. For each ciphertext block, starting at the _(n+1)th_ block we use the standard padding oracle attack to decrypt that block with an IV of all null bytes, effectively finding the raw decryption of that block. We then xor that decrypted block with the _nth_ plaintext block and set the _nth_ ciphertext block to the result of that operation.

**python-esque pseudocode:**
```python
plaintext = [b'test_plaintext_t', b'wo_blocks\x07\x07\x07\x07\x07\x07\x07']
ciphertext = [b'A'*16 for i in range(len(plaintext))]
for i in range(len(plaintext)-1, -1, -1):
     ciphertext[i] = xor(plaintext[i], decrypt(ciphertext[i+1], IV=b"\x00"*16))
```


[1]: Juliano Rizzo; Thai Duong (25 May 2010). [Practical Padding Oracle Attacks](http://www.usenix.org/event/woot10/tech/full_papers/Rizzo.pdf) (PDF). USENIX WOOT 2010.

q3st1on committed:
Add Hacky Encrypt mode

Added a hacky encryption mode. Duplicated the wrapper and callback functions so they can be modified for encryption mode if needed.

q3st1on committed:
Added test

Added a test for the encryption mode

q3st1on committed:
Bugfixes

Fixed bugs in library and tests. Now seem to be working

q3st1on committed:
Updated README

added demonstration of encryption mode to README

q3st1on committed:
fixed readme typo

elklepo commented on tests/test_padding_oracle.py:
```diff
@@ -14,6 +15,21 @@ def test_padding_oracle_basic():
+    plaintext = b'the quick brown fox jumps over the lazy dog'
+    ciphertext = cryptor.encrypt(plaintext)
+
+    padder = padding.PKCS7(128).padder()
```
This script already supports only PKCS7 padding - maybe its a good idea to shift the responsibility of padding the payload from user to the library?

elklepo commented on src/padding_oracle/legacy.py:
```diff
@@ -105,16 +120,54 @@ def plaintext_callback(plaintext: bytes):
+            else:
+                logger.error(result.message)
+
+    def plaintext_callback(plaintext: bytes):
```
`plaintext_callback` doesn't make much sense in encrypt mode because we generate ciphertext, not plaintext (see attached pic).
I'm also not sure if it makes sense to print the generated ciphertext in real time because it will not be readable.

I'd suggest to print the overall status of encryption, e.g. 
```
[INFO] 14/16 bytes encrypted in block 4/9
[INFO] 15/16 bytes encrypted in block 4/9
[INFO] 16/16 bytes encrypted in block 4/9
[INFO] 1/16 bytes encrypted in block 3/9
...
```

elklepo replied:
<img width="714" alt="image" src="https://github.com/djosix/padding_oracle.py/assets/12268051/cbf838c8-2de9-43a2-aefa-770d074564a4">

q3st1on replied:
Yeah I'll fix the printout tomorrow probably. Am in the middle of exams rn hence the slapped together code.

In terms of padding, I think adding a couple other padding schemes (say a few more pkcs and padding with a specified byte) would be good and then allowing a user supplied padding function to enable padding agnostic use of the library. That sound reasonable?

elklepo replied:
I think that until this lib supports only PKCS7 padding (by default in decrypt mode) it should behave the same in encrypt mode (PKCS7 by default).
Other types of padding (+ custom padding handlers) can be implemented later for both modes.
But the final word goes to the @djosix.

Good luck on your exams @q3st1on!

djosix replied:
For now, just sticking with PKCS#7 should be fine. 👌

elklepo reviewed (approved):
Verified encryption feature on one of the services vulnerable to Padding Oracle.

djosix commented on README.md:
````diff
@@ -63,6 +64,42 @@ plaintext = padding_oracle(
     num_threads = 16,
 )
 ```
+### Encryption
````
Nit: Needs an extra line above and below.

djosix commented on README.md:
````diff
@@ -63,6 +64,42 @@ plaintext = padding_oracle(
 )
 ```
+### Encryption
+To illustrate the usage, consider an example of forging a token for`https://vulnerable.website/api/?token=<.....>`:
````
Nit: Needs a space before the backtick character

djosix commented on src/padding_oracle/legacy.py:
```diff
@@ -58,33 +58,48 @@ def padding_oracle(ciphertext: Union[bytes, str],
                                    set (default: None)
         return_raw     (bool)      do not convert plaintext into bytes and
                                    unpad (default: False)
+        mode           (bool|str)  encrypt the payload (defaut: False/'decrypt')
```
`mode` is better to just be a `str` which accepts "encrypt" and "decrypt" (defaults to "decrypt") ☺️

q3st1on replied:
Fixed

djosix commented on src/padding_oracle/legacy.py:
```diff
@@ -105,16 +120,54 @@ def plaintext_callback(plaintext: bytes):
         plaintext = convert_to_bytes(plaintext, null_byte)
         plaintext = remove_padding(plaintext)
-
+    
```
flake8 will complain about these spaces!

djosix commented on tests/test_padding_oracle.py:
```diff
@@ -14,6 +15,21 @@ def test_padding_oracle_basic():
+                               cryptor.oracle, 4, null_byte=b'?', mode='encrypt')
+    decrypted = cryptor.decrypt(encrypted)
+
+    assert decrypted == plaintext
```
It will be better to write `assert encrypted == ciphertext` here 🤔

djosix commented on README.md:
```diff
@@ -1,3 +1,4 @@
+
```
Nit: This line is not needed :D

djosix commented on README.md:
```diff
@@ -30,7 +31,7 @@ Performance of padding_oracle.py was evaluated using [0x09] Cathub Party from ED
 
 ## How to Use 
-
+### Decryption
```
Nit: Needs a blank line above and below :D

djosix commented on src/padding_oracle/legacy.py:
```diff
@@ -33,21 +33,21 @@
     'padding_oracle',
 ]
 
-
```
Try this command 👍

```
flake8 --max-line-length=120
```

djosix commented on src/padding_oracle/legacy.py:
```diff
@@ -58,33 +58,48 @@ def padding_oracle(ciphertext: Union[bytes, str],
         raise TypeError('expect null with type bytes or str')
     if not len(null_byte) == 1:
         raise ValueError('null byte should have length of 1')
+    if not isinstance(mode, (bool, str)):
```
`mode` is better to just be a `str` which accepts "encrypt" and "decrypt" (defaults to "decrypt") ☺️

q3st1on replied:
Made it str only

djosix commented on src/padding_oracle/legacy.py:
```diff
@@ -58,33 +58,48 @@ def padding_oracle(ciphertext: Union[bytes, str],
 
+
+    # Does the user want the encryption routine
+    if (mode == 'encrypt') or (mode == True):
```
`mode` is better to just be a `str` which accepts "encrypt" and "decrypt" (defaults to "decrypt") ☺️

q3st1on replied:
made it str only

djosix reviewed (commented):
❤️

q3st1on commented:
Fixed the print out and other things a while ago but forgot to push changes lol. Will push asap

q3st1on committed:
Updated code

Fixed the printout to just update the user on progress (it does still print the final encrypted payload).
Added PKCS#7 padding by default to match the default decryption behaviour.
Changed mode to accept only a string.
I think I have fixed the various nit suggestions but may have missed some (or many tbh)
Actually remembered to push changes (wow incredible right)

q3st1on committed:
formatting issues

forgot to check formatting was ok, have fixed some issues with it. flake8 is not happy with the formatting

q3st1on commented:
I think everything mentioned here has been fixed. Would appreciate if someone else could have a quick read through double check that I didn't miss something.
